Add OpenJDK 17 images to OpenShift ImageStream templates.

jhuttana opened the pull request:
Could you please review?
These changes are done by branching out from 'release' branch.

jhuttana commented:
I have completed syntax checking with 'jq' tool for all the modified files. 
like : 

`$ jq . <json-file>`

I didnt see any error.

jhuttana commented:
As you mentioned below are the ways to test these changes

1. syntax checking by running through a tool like jq,
2.  importing them into an OpenShift cluster,
3. running the `make import`,` make verify` steps from a clone of https://github.com/openshift/library which has been adjusted to pull the templates from a development fork/branch.

Syntax check I have done. I am not sure about item 2. and 3.
As this is the first time I am trying testing part could you please help?

jmtd commented on templates/community-image-streams.json:
```diff
@@ -115,6 +115,59 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.3",
```
This segment describes ubi8/openjdk-17:1.3 but that version does not exist; the first version we have published is :1.10.

jmtd replied:
(and in the other files)

jmtd commented on templates/community-image-streams.json:
```diff
@@ -207,6 +260,46 @@
+                        }
+                    },
+                    {
+                        "name": "11",
```
this should be "17"

jmtd commented on templates/image-streams-ppc64le.json:
```diff
@@ -207,6 +207,85 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.0",
```
drop this section and the next (1.0 and 1.1)

jmtd commented:
> Syntax check I have done. I am not sure about item 2. and 3.
> As this is the first time I am trying testing part could you please help?

Sure no problem!

> 2. importing them into an OpenShift cluster,

I've logged into an OpenShift cluster (I'll send you details of a shared one internally if you like). Then:

```
$ oc create -n dotnet-java-playground -f https://raw.githubusercontent.com/jhuttana/openjdk/release_branch_OpenJDK17_images_to_OpenShift_ImageStream_templates/templates/community-image-streams.json
imagestream.image.openshift.io/ubi8-openjdk-8 created
imagestream.image.openshift.io/ubi8-openjdk-11 created
imagestream.image.openshift.io/ubi8-openjdk-17 created
imagestream.image.openshift.io/java created
```

That completed successfully.  Then I can inspect the objects it created:

```
;oc describe imagestream/ubi8-openjdk-17
Name:			ubi8-openjdk-17
Namespace:		dotnet-java-playground
Created:		About a minute ago
Labels:			<none>
Annotations:		openshift.io/display-name=Red Hat OpenJDK 17 (UBI8)
			openshift.io/image.dockerRepositoryCheck=2021-11-29T10:00:12Z
			openshift.io/provider-display-name=Red Hat, Inc.
Image Repository:	image-registry.openshift-image-registry.svc:5000/dotnet-java-playground/ubi8-openjdk-17
Image Lookup:		local=false
Unique Images:		1
Tags:			2

1.10
  tagged from registry.access.redhat.com/ubi8/openjdk-17:1.10
    prefer registry pullthrough when referencing this tag

  Build and run Java applications using Maven and OpenJDK 17 upon RHEL8.
  Tags: builder, java, openjdk, ubi8
  Example Repo: https://github.com/jboss-openshift/openshift-quickstarts

  * registry.access.redhat.com/ubi8/openjdk-17@sha256:9816c59fa8a2533c27960107a8bef181967d5af8903ac65a268c6a87407dd8fb
      About a minute ago

1.3
  tagged from registry.access.redhat.com/ubi8/openjdk-17:1.3
    prefer registry pullthrough when referencing this tag

  Build and run Java applications using Maven and OpenJDK 17 upon RHEL8.
  Tags: builder, java, openjdk, ubi8
  Example Repo: https://github.com/jboss-openshift/openshift-quickstarts

  ! error: Import failed (InternalError): Internal error occurred: registry.access.redhat.com/ubi8/openjdk-17:1.3: error parsing HTTP 404 response body: invalid character 'N' looking for beginning of value: "Not found\n"
      About a minute ago
```

That highlights that the metadata includes descriptions of non-existing image tags (1.3 for openjdk-17). but the :1.10 tag was successful.

> 3. running the make import, make verify steps from a clone of https://github.com/openshift/library which has been adjusted to pull the templates from a development fork/branch.

I haven't done this yet  but I'll do it and describe it on the next review round.

jhuttana commented:
@jmtd  Thank you for the review!! 
I have addressed the review comments. When I tried to build I faced a problem:
```
$ cekit --redhat --descriptor ubi8-openjdk-11.yaml build podman
2021-12-09 14:32:48,892 tools.py:244        INFO  You are running on known platform: Fedora Linux 35 (Workstation Edition)
2021-12-09 14:32:49,016 builder.py:72         INFO  Generating files for podman engine
2021-12-09 14:32:49,017 base.py:71         INFO  Initializing image descriptor...
2021-12-09 14:32:49,030 resource.py:196        INFO  Copying resource 'cct_module'...
2021-12-09 14:32:51,682 base.py:609        WARNING Module's 'jboss.container.openjdk.jdk' version 'openj9-8' does not follow PEP 440 versioning scheme (https://www.python.org/dev/peps/pep-0440), we suggest follow this versioning scheme in modules
2021-12-09 14:32:51,691 base.py:609        WARNING Module's 'jboss.container.openjdk.jdk' version 'openj9-11' does not follow PEP 440 versioning scheme (https://www.python.org/dev/peps/pep-0440), we suggest follow this versioning scheme in modules
2021-12-09 14:32:51,734 base.py:543        WARNING Module version not specified for 'jboss.container.prometheus' module, using '8.2' default version
2021-12-09 14:32:51,734 base.py:543        WARNING Module version not specified for 'jboss.container.jolokia' module, using '8.2' default version
2021-12-09 14:32:51,818 docker.py:26         INFO  Handling artifacts for docker...
2021-12-09 14:32:51,818 base.py:367        WARNING The image has ContentSets repositories specified, all other repositories are removed!
2021-12-09 14:32:51,818 base.py:411        INFO  Using Red Hat ODCS service to create composes
2021-12-09 14:32:59,760 resource.py:154        WARNING No value found for 'name' in '{"dest": "/tmp/artifacts/", "url": "http://odcs.engineering.redhat.com/composes/odcs-980883/compose/Temporary/odcs-980883.repo"}' artifact; using auto-generated value of 'odcs-980883.repo'
2021-12-09 14:32:59,760 resource.py:196        INFO  Copying resource 'odcs-980883.repo'...
2021-12-09 14:33:00,353 base.py:296        INFO  Rendering Dockerfile...
2021-12-09 14:33:00,416 base.py:346        INFO  Rendering help.md page from template /usr/lib/python3.10/site-packages/cekit/generator/../templates/help.jinja
2021-12-09 14:33:00,429 podman.py:48         INFO  Building container image...
STEP 1/108: FROM registry.access.redhat.com/ubi8/ubi-minimal
STEP 2/108: USER root
STEP 3/108: COPY repos/content_sets_odcs.repo     /etc/yum.repos.d/
STEP 4/108: COPY modules/jboss.container.microdnf-bz-workaround /tmp/scripts/jboss.container.microdnf-bz-workaround
STEP 5/108: USER root
STEP 6/108: RUN [ "sh", "-x", "/tmp/scripts/jboss.container.microdnf-bz-workaround/configure.sh" ]
+ mkdir -p /run/user/0
STEP 7/108: COPY modules/jboss.container.user /tmp/scripts/jboss.container.user
STEP 8/108: USER root
STEP 9/108: RUN microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y unzip tar rsync shadow-utils             && microdnf clean all             && rpm -q unzip tar rsync shadow-utils
error: (--setopt) Unable to handle: install_weak_deps=0
Error: error building at STEP "RUN microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y unzip tar rsync shadow-utils             && microdnf clean all             && rpm -q unzip tar rsync shadow-utils": error while running runtime: exit status 1
2021-12-09 14:33:02,279 cli.py:348        ERROR Image build failed, see logs above.

```
To verify whether this failure is only with my branch I tried build on 'release' branch. I see the same failure there also. 
Whereas, builds fine on 'develop' branch. A week back itself things were fine. Only change on my workstation is that OS upgraded to f35. And the cekit version is:
`
$ cekit --version
3.12.0
`
Further, when we push any changes to PR, Github with build and launch tests. That I didn't see on this branch. Could you please help me understand why is it so? (I assume its because we haven't setup necessary workflows under .github)

jhuttana commented:
Adding below flag to descriptor seems to solve the issue:
```
$ git diff
diff --git a/ubi8-openjdk-11.yaml b/ubi8-openjdk-11.yaml
index 62fd1f6..95c20a2 100644
--- a/ubi8-openjdk-11.yaml
+++ b/ubi8-openjdk-11.yaml
@@ -71,4 +71,5 @@ osbs:
 
 packages:
   manager: microdnf
+  manager_flags: --setopt=tsflags=nodocs
   content_sets_file: content_sets_rhel8.yml
```
I am not sure whether its correct way or not :)

jmtd commented:
We'll need to rebase this PR now that I merged ysekowitz's, and decide what we need to include.

There shouldn't be a need to rebuild the images as part of this PR (and the image sources in the `release` branch are by now quite out of date) but I'll try to help figure out the building issue:

> `+  manager_flags: --setopt=tsflags=nodocs`

What that effectively does is stop passing `--setopt=install_weak_deps=0` to `microdnf` which side-steps the problem, but I think recent versions of `microdnf` should accept that parameter (and we would like it: otherwise we will install extra packages in the images).

Can you check what version of the base image you have locally? e..g.

```
$ podman inspect -f '{{.Labels.version}}-{{.Labels.release}}' registry.access.redhat.com/ubi8/ubi-minimal
8.5-203
$ podman run --rm -ti registry.access.redhat.com/ubi8/ubi-minimal rpm -qa microdnf
microdnf-3.8.0-2.el8.x86_64
```

jhuttana commented:
> We'll need to rebase this PR now that I merged ysekowitz's, and decide what we need to include.

Sure
> 
> There shouldn't be a need to rebuild the images as part of this PR (and the image sources in the `release` branch are by now quite out of date) but I'll try to help figure out the building issue:
> 
> > `+  manager_flags: --setopt=tsflags=nodocs`
> 
> What that effectively does is stop passing `--setopt=install_weak_deps=0` to `microdnf` which side-steps the problem, but I think recent versions of `microdnf` should accept that parameter (and we would like it: otherwise we will install extra packages in the images).
> 
> Can you check what version of the base image you have locally? e..g.
> 
> ```
> $ podman inspect -f '{{.Labels.version}}-{{.Labels.release}}' registry.access.redhat.com/ubi8/ubi-minimal
> 8.5-203
> $ podman run --rm -ti registry.access.redhat.com/ubi8/ubi-minimal rpm -qa microdnf
> microdnf-3.8.0-2.el8.x86_64
> ```

Above steps helped me to know that I had old registry and hence old version of microdnf. Now I have pulled the registry again and matches the same values as yours.

jhuttana commented:
@jmtd I have rebased this branch and resolved the conflicts. Could you please take a look?

pestapever commented:
News? Once merged, an openjdk17rhel image will be created on the redhat catalog, such as https://catalog.redhat.com/software/containers/openjdk/openjdk-11-rhel7/5bf57185dd19c775cddc4ce5, correct?

jerboaa commented:
> News? Once merged, an openjdk17rhel image will be created on the redhat catalog, such as https://catalog.redhat.com/software/containers/openjdk/openjdk-11-rhel7/5bf57185dd19c775cddc4ce5, correct?

The JDK 17 image is already in the catalog. It's UBI8+ only, though. See:
https://catalog.redhat.com/software/containers/ubi8/openjdk-17/618bdbf34ae3739687568813

This PR is about an image stream pointing to those images.

jerboaa commented on templates/image-streams-s390x.json:
```diff
@@ -210,6 +210,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.0"
```
There is no such image. JDK 17 is RHEL 8/UBI8+ only.

jerboaa commented on templates/image-streams-s390x.json:
```diff
@@ -210,6 +210,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.1"
```
Same here. No RHEL 7 JDK 17 image.

jerboaa commented on templates/image-streams-s390x.json:
```diff
@@ -210,6 +210,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.10"
```
Should be pointing to `registry.access.redhat.com/ubi8/openjdk-17:1.10`

jhuttana replied:
This is also rhel7 related entry. As you mentioned there is no RHEL7 JDK17, I think we can delete this.

jerboaa commented on templates/image-streams-s390x.json:
```diff
@@ -324,6 +403,63 @@
+                        }
+                    },
+                    {
+                        "name": "1.10",
```
Duplicate tag? `1.10` listed twice?

jhuttana replied:
Yes :) Bad copy paste ;)

jerboaa commented on templates/image-streams-s390x.json:
```diff
@@ -631,6 +767,46 @@
+                        }
+                    },
+                    {
+                        "name": "openjdk-17-el7",
```
No such image. Please remove.

jerboaa commented on templates/image-streams.json:
```diff
@@ -462,6 +598,44 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel8:1.0"
```
There is only an UBI8 image. Please remove.

jerboaa reviewed (changes requested):
@jhuttana There are some references to images that don't exist. e.g. `openjdk/openjdk-17-rhel7` and `openjdk/openjdk-17-rhel8`. Please remove those references. The initial tag published was `1.10` AFAIK, no older tags should be added.

pestapever commented:
I used to create a Quarkus app for openshift S2I strategy with command "oc new-app openjdk-11-rhel7 ...", with ubi I got problems and build errors.. I'm waiting for a jdk17rhel8 to upgrade java version with new applications in Quarkus..

jmtd commented:
> I used to create a Quarkus app for openshift S2I strategy with command "oc new-app openjdk-11-rhel7 ...", with ubi I got problems and build errors.. I'm waiting for a jdk17rhel8 to upgrade java version with new applications in Quarkus..

Redhat are not producing a jdk17rhel8 container, other than ubi8/openjdk-17 and ubi8/openjdk-17-runtime. If you can share details of the problems you had with the UBI images we might be able to help you resolve those.

jhuttana commented:
@jerboaa I have addressed the review comments. Could you please take a look?

jerboaa commented on templates/image-streams.json:
```diff
@@ -310,6 +310,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.10"
```
Still showing rhel7, which doesn't exist.

jerboaa commented on templates/image-streams.json:
```diff
@@ -310,6 +310,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.1"
```
Not a valid image.

jerboaa commented on templates/image-streams.json:
```diff
@@ -310,6 +310,85 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/openjdk/openjdk-17-rhel7:1.0"
```
Same, no valid image.

jhuttana commented:
@jerboaa Could you please take a look?

jerboaa reviewed (approved):
Looks good. Please test it on an OpenShift test instance.

jhuttana commented:
> Looks good. Please test it on an OpenShift test instance.

ok thank you!

jmtd commented on templates/image-streams.json:
```diff
@@ -461,7 +518,7 @@
         {
             "kind": "ImageStream",
-            "apiVersion": "image.openshift.io/v1",
+            "apiVersion": "v1",
```
leave this line as it was please

jmtd commented on templates/image-streams.json:
```diff
@@ -424,6 +424,63 @@
+                            "name": "registry.access.redhat.com/ubi8/openjdk-17:1.10"
+                        }
+                    },
+                    {
```
this section (lines 459-477) appears to be a duplicate item to the previous one and should be removed

jmtd commented on templates/image-streams.json:
```diff
@@ -309,7 +309,7 @@
         {
             "kind": "ImageStream",
-            "apiVersion": "image.openshift.io/v1",
+            "apiVersion": "v1",
```
please undo this change

jmtd commented on templates/image-streams-s390x.json:
```diff
@@ -209,7 +209,7 @@
         {
             "kind": "ImageStream",
-            "apiVersion": "image.openshift.io/v1",
+            "apiVersion": "v1",
```
please undo this change

jmtd commented on templates/image-streams-ppc64le.json:
```diff
@@ -324,6 +324,63 @@
+                            "name": "registry.access.redhat.com/ubi8/openjdk-17:1.10"
+                        }
+                    },
+                    {
```
this section (lines 359-377) is a duplicate of the previous item and should be removed

jmtd commented on templates/image-streams-aarch64.json:
```diff
@@ -126,6 +126,63 @@
+                            "name": "registry.access.redhat.com/ubi8/openjdk-17:1.10"
+                        }
+                    },
+                    {
```
this section (lines 161-179) is a duplicate of the previous item and should be removed

jmtd commented on templates/community-image-streams.json:
```diff
@@ -118,6 +118,59 @@
+                            "name": "registry.access.redhat.com/ubi8/openjdk-17:1.10"
+                        }
+                    },
+                    {
```
this item (lines 149-167) is a duplicate of the previous item and should be removed

jmtd commented:
We are very soon releasing version 1.11 of the OpenJDK 17 images. I would suggest putting 1.11 in as the version instead of 1.10 (we could do it as well as, but since we have never had the openjdk 17 images in this metadata, there's no need to add an old version). Just substitute 1.10 for 1.11 and we'll merge once the PR is ready and 1.11 images are available. Thanks!

jhuttana commented:
@jmtd Could you please take a look at recent changes addressing review comments?

jhuttana commented:
> We are very soon releasing version 1.11 of the OpenJDK 17 images. I would suggest putting 1.11 in as the version instead of 1.10 (we could do it as well as, but since we have never had the openjdk 17 images in this metadata, there's no need to add an old version). Just substitute 1.10 for 1.11 and we'll merge once the PR is ready and 1.11 images are available. Thanks!


I have changed 1.10->1.11 for OpenJDK 17 images. Same changes for 11 and 8 also?

jerboaa commented:
> Same changes for 11 and 8 also?

No. We've released image streams for those in the past.

jhuttana commented:
> > Same changes for 11 and 8 also?
> 
> No. We've released image streams for those in the past.


Ok. Thanks for the clarification :)

jmtd commented on templates/community-image-streams.json:
```diff
@@ -118,6 +118,40 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.10",
```
please change to 1.11

jmtd commented on templates/community-image-streams.json:
```diff
@@ -118,6 +118,40 @@
+                            "tags": "builder,java,openjdk,ubi8",
+                            "sampleRepo": "https://github.com/jboss-openshift/openshift-quickstarts",
+                            "sampleContextDir": "undertow-servlet",
+                            "version": "1.10"
```
please change to 1.11

jmtd commented on templates/image-streams-aarch64.json:
```diff
@@ -126,6 +126,44 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.10",
```
please change to 1.11

jmtd commented on templates/image-streams-aarch64.json:
```diff
@@ -126,6 +126,44 @@
+                            "version": "1.10"
+                        },
+                        "referencePolicy": {
+                            "type": "Local"
```
please change to 1.11

jmtd commented on templates/image-streams-ppc64le.json:
```diff
@@ -324,6 +324,44 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.10",
```
please change to 1.11

jmtd commented on templates/image-streams-ppc64le.json:
```diff
@@ -324,6 +324,44 @@
+                            "tags": "builder,java,openjdk,ubi8,hidden",
+                            "sampleRepo": "https://github.com/jboss-openshift/openshift-quickstarts",
+                            "sampleContextDir": "undertow-servlet",
+                            "version": "1.10"
```
please change to 1.11

jmtd commented on templates/image-streams-s390x.json:
```diff
@@ -324,6 +324,44 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.10",
```
please change to 1.11

jmtd commented on templates/image-streams-s390x.json:
```diff
@@ -324,6 +324,44 @@
+                            "tags": "builder,java,openjdk,ubi8,hidden",
+                            "sampleRepo": "https://github.com/jboss-openshift/openshift-quickstarts",
+                            "sampleContextDir": "undertow-servlet",
+                            "version": "1.10"
```
please change to 1.11

jmtd commented on templates/image-streams.json:
```diff
@@ -424,6 +424,44 @@
+            "spec": {
+                "tags": [
+                    {
+                        "name": "1.10",
```
please change to 1.11

jmtd commented on templates/image-streams.json:
```diff
@@ -424,6 +424,44 @@
+                            "tags": "builder,java,openjdk,ubi8,hidden",
+                            "sampleRepo": "https://github.com/jboss-openshift/openshift-quickstarts",
+                            "sampleContextDir": "undertow-servlet",
+                            "version": "1.10"
```
please change to 1.11

jhuttana commented:
@jmtd Thanks for your patience :) 
Could you please take a look?

jmtd commented on templates/community-image-streams.json:
```diff
@@ -118,6 +118,40 @@
+        },
+        {
+            "kind": "ImageStream",
+            "apiVersion": "v1",
```
please change this value to image.openshift.io/v1

jmtd commented on templates/image-streams-aarch64.json:
```diff
@@ -126,6 +126,44 @@
+        },
+        {
+            "kind": "ImageStream",
+            "apiVersion": "v1",
```
and here

jmtd commented on templates/image-streams-ppc64le.json:
```diff
@@ -324,6 +324,44 @@
+        },
+        {
+            "kind": "ImageStream",
+            "apiVersion": "v1",
```
and here

jmtd commented on templates/image-streams-s390x.json:
```diff
@@ -324,6 +324,44 @@
+        },
+        {
+            "kind": "ImageStream",
+            "apiVersion": "v1",
```
andhere

jmtd commented on templates/image-streams-s390x.json:
```diff
@@ -631,6 +669,26 @@
+                        },
+                        "from": {
+                            "kind": "DockerImage",
+                            "name": "registry.redhat.io/ubi8/openjdk-17:latest"
```
please change the host to registry.access.redhat.com

jmtd commented on templates/image-streams.json:
```diff
@@ -424,6 +424,44 @@
+        },
+        {
+            "kind": "ImageStream",
+            "apiVersion": "v1",
```
and here

jmtd commented:
> @jmtd Thanks for your patience :)

No problem: a couple more minor nits and then we're just waiting on the new images to ship!

jhuttana commented:
@jmtd Could you take  a look at recent changes?

jhuttana commented:
@jerboaa I have tested these changes on openshift test instance and is successful.
@jmtd Thank you for guiding me on that.
